wkcuber/api/Layer.py

```diff
@@ -0,0 +1,142 @@
+from shutil import rmtree
+from os import mkdir
+from os.path import join
+import logging
+from wkw import wkw
+
+from wkcuber.api.MagDataset import MagDataset, WKMagDataset, TiffMagDataset
+from wkcuber.mag import Mag
+
+
+class Layer:
+    def __init__(self, name, dataset, dtype, num_channels):
+        self.name = name
+        self.dataset = dataset
+        self.dtype = dtype
+        self.num_channels = num_channels
+        self.mags = {}
+
+        full_path = join(dataset.path, name)
+        try:
+            mkdir(full_path)
+        except OSError:
+            logging.info("Creation of Layer {} failed".format(full_path))
+
+    def get_mag(self, mag) -> MagDataset:
+        if mag not in self.mags.keys():
+            raise IndexError("The mag {} is not a mag of this layer".format(mag))
+        return self.mags[mag]
+
+    def delete_mag(self, mag):
+        if mag not in self.mags.keys():
+            raise IndexError(
+                "Deleting mag {} failed. There is no mag with this name".format(mag)
+            )
+
+        del self.mags[mag]
+        self.dataset.properties.delete_mag(self.name, mag)
+        # delete files on disk
+        full_path = join(self.dataset.path, self.name, mag)
+        rmtree(full_path)
+
+    def __create_dir_for_mag__(self, mag):
+        full_path = join(self.dataset.path, self.name, mag)
+
+        try:
+            mkdir(full_path)
+        except OSError:
+            logging.info("Creation of MagDataset {} failed".format(full_path))
+
+    def __assert_mag_does_not_exist_yet__(self, mag):
+        if mag in self.mags.keys():
+            raise IndexError(
+                "Adding mag {} failed. There is already a mag with this name".format(
+                    mag
+                )
+            )
+
+
+class WKLayer(Layer):
+    def add_mag(self, mag, block_len=32, file_len=32, block_type=1):
+        # normalize the name of the mag
+        mag = Mag(mag).to_layer_name()
+
+        self.__assert_mag_does_not_exist_yet__(mag)
+        self.__create_dir_for_mag__(mag)
+
+        self.mags[mag] = WKMagDataset.create(self, mag, block_len, file_len, block_type)
+        self.dataset.properties.add_mag(self.name, mag, block_len * file_len)
+
+        return self.mags[mag]
+
+    def get_or_add_mag(self, mag, block_len=32, file_len=32, block_type=1):
+        # normalize the name of the mag
+        mag = Mag(mag).to_layer_name()
+
+        if mag in self.mags.keys():
+            assert self.mags[mag].header.block_len == block_len, (
+                "Cannot get_or_add_mag: The mag %s already exists, but the block lengths do not match"
```

Comment:
See my other feedback regarding the assertion messages.

Reply:
What happens when I want to `get` a mag (from which I know that it exists), but I don't know with which block_len etc. it was created? Then, I don't have a possibility to use that method, right (since the default block_len value might not match the existing mag)?
I think, you probably need to adapt the default values to `None` and map that to actual default values *only when constructing a new mag*. The same is probably true for all `get_or_create` methods.